src/hdmf/common/resources.py

```diff
@@ -293,41 +325,74 @@ def get_resource(self, **kwargs):
-            {'name': 'entity_uri', 'type': str, 'doc': 'the URI for the identifier at the resource', 'default': None})
+            {'name': 'entity_uri', 'type': str, 'doc': 'the URI for the identifier at the resource', 'default': None}
+            )
     def add_ref(self, **kwargs):
         """
         Add information about an external reference used in this file.
 
         It is possible to use the same name of the key to refer to different resources
-        so long as the name of the key is not used within the same object and field.
-        This method does not support such functionality by default. The different
-        keys must be added separately using _add_key and passed to the *key* argument
-        in separate calls of this method. If a resource with the same name already
-        exists, then it will be used and the resource_uri will be ignored.
+        so long as the name of the key is not used within the same object, relative_path, and
+        field combination. This method does not support such functionality by default.
         """
+        ###############################################################
         container = kwargs['container']
-        field = kwargs['field']
+        attribute = kwargs['attribute']
         key = kwargs['key']
+        field = kwargs['field']
         entity_id = kwargs['entity_id']
         entity_uri = kwargs['entity_uri']
         add_entity = False
 
-        object_field = self._check_object_field(container, field)
+        if attribute is None:  # Trivial Case
+            relative_path = ''
+            object_field = self._check_object_field(container, relative_path, field)
+        else:  # DataType Attribute Case
+            attribute_object = getattr(container, attribute)  # returns attribute object
+            if isinstance(attribute_object, AbstractContainer):
+                relative_path = ''
+                object_field = self._check_object_field(attribute_object, relative_path, field)
+            else:  # Non-DataType Attribute Case:
+                obj_mapper = self.type_map.get_map(container)
+                spec = obj_mapper.get_attr_spec(attr_name=attribute)
+                parent_spec = spec.parent  # return the parent spec of the attribute
+                if parent_spec.data_type is None:
+                    while parent_spec.data_type is None:
+                        parent_spec = parent_spec.parent  # find the closest parent with a data_type
+                    parent_cls = self.type_map.get_dt_container_cls(data_type=parent_spec.data_type, autogen=False)
+                    if isinstance(container, parent_cls):
+                        parent_id = container.object_id
+                        # We need to get the path of the spec for relative_path
+                        absolute_path = spec.path
+                        relative_path = re.sub("^.+?(?="+container.data_type+")", "", absolute_path)
```

Comment:
What does this relative expression do? Please add a comment. Also note that `pathlib` is a handy standard library for path manipulation that may be useful here.